src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5297,13 +5297,15 @@ public static unsafe Task<TResult> FromResult<TResult>(TResult result)
             }
             else if (!RuntimeHelpers.IsReferenceOrContainsReferences<TResult>())
             {
-                // For other value types, we special-case default(TResult) if we can easily compare bit patterns to default/0.
+                // For other value types, we special-case default(TResult) if we can efficiently compare bit patterns to default/0,
+                // which means any value type that's 1, 2, 4, 8, or 16 bytes in size and that doesn't contain references.
                 // We don't need to go through the equality operator of the TResult because we cached a task for default(TResult),
                 // so we only need to confirm that this TResult has the same bits as default(TResult).
                 if ((sizeof(TResult) == sizeof(byte) && *(byte*)&result == default(byte)) ||
                     (sizeof(TResult) == sizeof(ushort) && *(ushort*)&result == default(ushort)) ||
                     (sizeof(TResult) == sizeof(uint) && *(uint*)&result == default) ||
-                    (sizeof(TResult) == sizeof(ulong) && *(ulong*)&result == default))
+                    (sizeof(TResult) == sizeof(ulong) && *(ulong*)&result == default) ||
+                    (sizeof(TResult) == sizeof(UInt128) && *(UInt128*)&result == default))
```

Comment:
For `UInt128` the `==` operator isn't vectorized, it does two comparisons (lower and upper).
So perf-wise -- if it matters -- `Guid` should be better, as that `==` is vectorized, and has only one comparison in scalar code (after the move mask).

Or Int128/UInt128 gets a vectorized version of equality.
(I can look into this, but not in the before two weaks)

Reply:
FWIW, I tried using a vector operation explicitly here and it was worse, in particular when the value wasn't 0.  If you could try as well, that'd be great; it's possible I was holding it wrong. This can also be done as a follow-up.

Reply:
I'll give it a try -- either here directly or via optimization of (U)Int128 equals (but a said a bit later).